lib/topological_inventory/orchestrator/metric_scaler.rb

```diff
@@ -0,0 +1,186 @@
+require "manageiq-loggers"
+
+module TopologicalInventory
+  module Orchestrator
+    class MetricScaler
+      attr_reader :logger
+
+      def initialize(logger = nil)
+        @logger = logger || ManageIQ::Loggers::CloudWatch.new
+        @cache  = {}
+      end
+
+      def run
+        loop do
+          run_once
+
+          sleep 10
+        end
+      end
+
+      def run_once
+        logger.info("#{self.class.name}##{__method__} Starting...")
+        dc_names = object_manager.get_deployment_configs("metric_scaler_enabled=true").collect { |dc| dc.metadata.name }
+
+        # newly_configured
+        (dc_names - @cache.keys).each { |name| @cache[name] = Watcher.new(name, logger).tap(&:start) }
+        # no_longer_configured
+        (@cache.keys - dc_names).each { |name| @cache.delete(name).stop }
+        # currently configured
+        dc_names.each                 { |name| @cache[name].scale_to_desired_replicas }
+
+        logger.info("#{self.class.name}##{__method__} Complete")
+      end
+
+      private
+
+      class Watcher
+        attr_reader :deployment_config, :deployment_config_name, :logger, :thread
+
+        def initialize(deployment_config_name, logger)
+          @deployment_config_name = deployment_config_name
+          @logger = logger
+          logger.info("Metrics scaling enabled for #{deployment_config_name}")
+          configure
+        end
+
+        def configured?
+          @current_metric_name && @max_metric_name && @max_replicas && @min_replicas && @target_usage_pct && @scale_threshold_pct
+        end
+
+        def start
+          @thread = Thread.new do
+            logger.info("Watcher thread for #{deployment_config_name} starting")
+            loop do
+              configure
+              break unless configured?
+
+              60.times do # Collect metrics for ~1 minute then check for config changes
+                moving_usage << percent_usage_from_metrics
+                sleep 1
+              end
+            end
+            logger.info("Watcher thread for #{deployment_config_name} stopping")
+          end
+        end
+
+        def stop
+          @thread.exit
+          @thread.join
+        end
+
+        def scale_to_desired_replicas
+          return unless configured?
+
+          desired_count = desired_replicas
+
+          return if desired_count == deployment_config.spec.replicas # already at max or minimum
+
+          logger.info("Scaling #{deployment_config_name} to #{desired_count} replicas")
+          object_manager.scale(deployment_config_name, desired_count)
+
+          # Wait for scaling to complete in Openshift
+          sleep(1) until pod_ips.length == desired_count
+          logger.info("Scaling #{deployment_config_name} complete")
+        end
+
+        private
+
+        def moving_usage
```

Comment:
This name is ... not great. Maybe `metric_list`? The name just doesn't convey that it's an array so the line `moving_usage.average` sounds weird.